Close enhance identification of IINA build, #4383

low-batt opened the pull request:
This commit will:
- Add Beta and Nightly Xcode build configurations to the Xcode project
- Add Beta.xcconfig and Nightly.xcconfig build configuration files for the new build configuration
- Add two new icons, AppIconBeta and AppIconNightly
- Use ASSETCATALOG_COMPILER_APPICON_NAME to select the correct application icon
- Add a key com.colliderli.iina.build.configuration to Info.plist to allow IINA to determine at runtime the build configuration used to generate the executable
- Add a property buildConfiguration to InfoDictionary that exposes the raw value in Info.plist
- Add a BuildType enumeration that corresponds to the build configurations in the Xcode project
- Add buildType, buildTypeIdentifier and isRelease properties to InfoDictionary to make it easy for other code to access the build configuration

This commit is only a start on the changes needed to implement #4383. This commit only contains the changes to Xcode needed to use different application icons for Beta and Nightly builds and to allow IINA to know what type of build it is running. The new application icons have been populated with document icons for testing. These must be replaced with the appropriate icons once they have been generated. The commit does not contain any changes to the welcome window or about window. Those changes still have to be implemented.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #4383.

---

**Description:**

low-batt committed:
Close enhance identification of IINA build, #4383

This commit will:
- Add Beta and Nightly Xcode build configurations to the Xcode project
- Add Beta.xcconfig and Nightly.xcconfig build configuration files for
  the new build configuration
- Add two new icons, AppIconBeta and AppIconNightly
- Use ASSETCATALOG_COMPILER_APPICON_NAME to select the correct
  application icon
- Add a key com.colliderli.iina.build.configuration to Info.plist to
  allow IINA to determine at runtime the build configuration used to
  generate the executable
- Add a property buildConfiguration to InfoDictionary that exposes the
  raw value in Info.plist
- Add a BuildType enumeration that corresponds to the build
  configurations in the Xcode project
- Add buildType, buildTypeIdentifier and isRelease properties to
  InfoDictionary to make it easy for other code to access the build
  configuration

This commit is only a start on the changes needed to implement #4383.
This commit only contains the changes to Xcode needed to use different
application icons for Beta and Nightly builds and to allow IINA to know
what type of build it is running. The new application icons have been
populated with document icons for testing. These must be replaced with
the appropriate icons once they have been generated. The commit does not
contain any changes to the welcome window or about window. Those changes
still have to be implemented.

low-batt commented:
This draft PR only contains the changes to Xcode needed to use different application icons for Beta and Nightly builds and to allow IINA to know what type of build it is running.

No changes have been made to the welcome and about windows. They still need to be updated to indicate the type of build.

Two new application icon sets, `AppIconBeta` and `AppIconNightly` were added as assets. They have been populated with document icons for testing. These must be replaced with the appropriate icons once they have been generated.

The log file has been updated to indicate the build type:
```text
17:50:49.168 [iina][d] IINA 1.3.1 Build 133 DEBUG
```

Details on the Xcode changes follow…

Two new configurations, `Beta` and `Nightly` were added by opening the IINA project, selecting the `Info` tab, and using the `+` button in the `Configurations` section and selecting `Duplicate "Release" Configuration` in the popup. The instructions in the Apple article [Adding a build configuration file to your project](https://developer.apple.com/documentation/xcode/adding-a-build-configuration-file-to-your-project) were then followed to add `Beta.xcconfig` and `Nightly.xcconfig` configuration files.

The configuration files make use of the following standard Xcode [Build settings](https://developer.apple.com/documentation/xcode/build-settings-reference):

- [Primary App Icon Set Name](https://developer.apple.com/documentation/xcode/build-settings-reference#Primary-App-Icon-Set-Name)
    > **Setting name:** ASSETCATALOG_COMPILER_APPICON_NAME
    > 
    > Name of an app icon set for the target’s default app icon. The contents will be merged into the Info.plist.

- [CONFIGURATION](https://developer.apple.com/documentation/xcode/build-settings-reference#CONFIGURATION)
    >**Setting name:** CONFIGURATION
    >
    > Identifies the build configuration, such as Debug or Release, that the target uses to generate the product.

A new entry was then added to the `Info.plist` file following the instructions in [Edit property lists](https://help.apple.com/xcode/mac/current/#/dev3f399a2a6). Prefixing custom keys with the bundle identifier is recommended by Apple so the new key is named `com.colliderli.iina.build.configuration`. It is assigned the value of `$(CONFIGURATION)`. This enables IINA at runtime to determine how it was built.

Properties were added to `InfoDictionary` to make it easy for other code to determine the build type.

uiryuu committed:
Properly localize the initial window

- The initial window now shows information based on build type
- Added 2 new colors for nightly and debug

low-batt commented on iina/Base.lproj/Localizable.strings:
```diff
@@ -21,6 +21,13 @@
+"initial.announcement" = "You are using a %@ version of IINA";
+"initial.beta.desc" = "Uncheck \"receive beta updates\" from Settings > General, and install a stable release from our <a href=\"https://iina.io\">official website</a>.";
+"initial.nightly.desc" = "Nightly is not recommanded to be used as daily drive. Currently nightly builds do not support auto update. Please manually download newer builds from our <a href=\"https://nightly.iina.io\">nightly website</a>.";
+"initial.debug.desc" = "Debug builds are usually not optimized for performance, so it is not recommanded to be used as daily drive. Debug builds do not support auto update.";
```
typo recommanded should be recommended

low-batt commented on iina/Base.lproj/Localizable.strings:
```diff
@@ -21,6 +21,13 @@
+"initial.beta.desc" = "Uncheck \"receive beta updates\" from Settings > General, and install a stable release from our <a href=\"https://iina.io\">official website</a>.";
+"initial.nightly.desc" = "Nightly is not recommanded to be used as daily drive. Currently nightly builds do not support auto update. Please manually download newer builds from our <a href=\"https://nightly.iina.io\">nightly website</a>.";
+"initial.debug.desc" = "Debug builds are usually not optimized for performance, so it is not recommanded to be used as daily drive. Debug builds do not support auto update.";
+"initial.bug_report" = "Report to us via <a href=\"https://github.com/iina/iina/issues\">GitHub issues</a>.";
```
Could shorten this to just "Report via"

low-batt commented on iina/InitialWindowController.swift:
```diff
@@ -64,6 +64,20 @@ fileprivate extension NSColor {
   }()
+  static let initialWindowNightlyLabel: NSColor = {
+    if #available(macOS 10.14, *) {
+      return NSColor(named: .initialWindowNightlyLabel)!
```
I was unable to test Nightly as a run blows up here with a nil.

uiryuu replied:
My fault, I unstaged too many changes.

low-batt commented on iina/InitialWindowController.swift:
```diff
@@ -64,6 +64,20 @@ fileprivate extension NSColor {
+  }()
+  static let initialWindowDebugLabel: NSColor = {
+    if #available(macOS 10.14, *) {
+      return NSColor(named: .initialWindowDebugLabel)!
```
I was unable to test Debug as a run blows up here with a nil.

low-batt commented on iina/en.lproj/Localizable.strings:
```diff
@@ -21,6 +21,13 @@
+// Initial Window
+"initial.announcement" = "You are using a %@ version of IINA";
+"initial.beta.desc" = "Uncheck \"receive beta updates\" from Settings > General, and install a stable release from our <a href=\"https://iina.io\">official website</a>.";
+"initial.nightly.desc" = "Nightly is not recommanded to be used as daily drive. Currently nightly builds do not support auto update. Please manually download newer builds from our <a href=\"https://nightly.iina.io\">nightly website</a>.";
```
recommended

low-batt commented on iina/InitialWindowController.swift:
```diff
@@ -453,16 +482,29 @@ class InitialWindowViewActionButton: NSView {
+      break
+    }
+
+    announcementLabel.stringValue = String(format: NSLocalizedString("initial.announcement", comment: "Version annoucnement"), buildType.rawValue)
```
typo "Version annoucnement" should be "Version announcement"

uiryuu committed:
Added missing colors

uiryuu committed:
Remove debug as a build type

- fixed some typos

low-batt commented on Configs/Debug.xcconfig:
```diff
@@ -8,8 +8,8 @@
-ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon
-CONFIGURATION = Debug
+ASSETCATALOG_COMPILER_APPICON_NAME = AppIconNightly
+CONFIGURATION = Nightly
```
I am so not understanding this change. We want the nightly build to be a release build, yes? The debug builds are pretty much for developers. Developers need to be able to distinguish them from nightly builds.

low-batt commented on iina/InfoDictionary.swift:
```diff
@@ -83,7 +87,6 @@ struct InfoDictionary {
   enum BuildType: String, CustomStringConvertible {
     case beta = "Beta"
-    case debug = "Debug"
     case nightly = "Nightly"
```
This change makes the above comment incorrect. Xcode has a Debug build configuration, so you would expect it to be here. In other words this is no longer a direct mapping to Xcode configurations as the comment would have you believe. This is confusing and definitely needs a comment to explain the thinking. Possibly whatever we are trying to achieve is being done at too low of a level. In other words this low level code should do the simple thing and map to the Xcode configurations. That is easy and straight forward for people to understand. Higher level code can then treat the debug configuration differently, like mapping it to nightly.

low-batt commented:
I was not able to test and see what the latest changes look like because something about them has really broken Xcode. It seems to only be building `iina-cli` so when it goes to start IINA this is all the Xcode console shows:
```text
Cannot find IINA binary. This command line tool only works in IINA.app bundle.
Program ended with exit code: 1
```

low-batt commented:
I figured out that the changes switched the scheme that was being built on me to `iina-cli`. If you bring up `Manage Schemes…` it is possible to change the order of the schemes. I think the IINA schemes should come first.

I was able to build and run `beta` and `nightly`. Something has caused the configuration to disappear from the log file:
```text
14:40:58.631 [iina][d] IINA 1.3.1 Build 133
```

I was unable to build `debug`:

<img width="1789" alt="Screenshot 2023-04-29 at 2 39 17 PM" src="https://user-images.githubusercontent.com/86170219/235319482-570469a8-5b6f-444e-bdcd-0ee04ee7018f.png">



Lots of these errors:
```text
info.plist found in '/Users/low-batt/Library/Developer/Xcode/DerivedData/iina-gwhhvjgxybcchtdcwvmukizbfeip/SourcePackages/artifacts/sparkle/Sparkle/sparkle.app/Contents' but failed to parse: keyNotFound(CodingKeys(stringValue: "AvailableLibraries", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"AvailableLibraries\", intValue: nil) (\"AvailableLibraries\").", underlyingError: nil))
```

saagarjha commented:
@low-batt has mentioned some of this already but this is the hierarchy I expect:

* Debug configuration
	* Debug.xcconfig
		* iina target (local build)
* Release configuration
	* Release.xcconfig
		* iina target (local build)
* Beta configuration
	* Release.xcconfig
		* iina target (beta build)
* Nightly configuration
	* Release.xcconfig
		* iina target (nightly build)
* Production configuration
	* Release.xcconfig
		* iina target (production build)

Debug and Release exist for local development and profiling. They currently use what is called "iina.xcconfig". Beta, Nightly, and Production are identical from a build layout perspective; they only exist to ship to users and exist with Release.xcconfig applied. The target-level xcconfig is basically iina.xcconfig, with a specific app icon set. How you want to do this is up to you. One options is this:

### iina.xcconfig (not used by any targets)
```
[Current contents of iina.xcconfig save for the app icon]
```

### iina-local.xcconfig (used by Debug, Release)
```
#include "iina.xcconfig"

ASSETCATALOG_COMPILER_APPICON_NAME = LocalAppIcon
```

### iina-beta.xcconfig (used by Beta)
```
#include "iina.xcconfig"

ASSETCATALOG_COMPILER_APPICON_NAME = BetaAppIcon
```

…and so on for the Nightly and Production builds.

uiryuu commented:
@saagarjha Any updates? Or should I continue to implement this PR?

saagarjha commented:
Oh I just self-assigned myself to review the updated version before you merged it

uiryuu committed:
Add back Debug config

- unshared xcschemes
- added newly designed app icons
- updated the colors

low-batt commented:
With the latest commit, building and running `iina` crashes out while trying to force unwrap the debug label:
```swift
  static let initialWindowDebugLabel: NSColor = {
    if #available(macOS 10.14, *) {
      return NSColor(named: .initialWindowDebugLabel)!
```

uiryuu committed:
Fix uppercase I in "Window"

uiryuu committed:
Update the about window

- Make the branch label clickable (link to the commit on GitHub)
- Use unbordered NSButton instead of NSTextField

uiryuu commented:
@low-batt I didn't show the "BETA" and "NIGHLY" label in the about window because we already have the icon which already clearly shows which version we are on, so showing another label looks redundant.

low-batt commented:
I tested all the different kinds of builds.

Debug builds are properly identified in the log file:
```text
11:54:33.622 [iina][d] IINA 1.3.1 Build 133 DEBUG
```

Nightly and Beta builds are no longer identified:
```text
11:56:33.547 [iina][d] IINA 1.3.1 Build 133
```
This is because isRelease is true for Nightly and Beta builds.
It used to key off the buildType:
```swift
var isRelease: Bool { buildType == .release }
```

But was changed to:
```swift
  #if DEBUG
  let isRelease = false
  #else
  let isRelease = true
  #endif
```

The commit number on the welcome window still needs to be changed to be a link to GitHub.
Same for the about window

Otherwise looking good!

uiryuu committed:
Fix for logger

uiryuu commented:
@low-batt I've updated to make logger pickup the release type.

> The commit number on the welcome window still needs to be changed to be a link to GitHub. Same for the about window

The commit number in the about window is clickable. I don't think it's necessary to make the commit number in the welcome window clickable. If we really want to, that need a better design (other than just make it a NSButton). So I'd leave it as is for this PR.

low-batt commented on iina/InfoDictionary.swift:
```diff
@@ -44,14 +50,57 @@ struct InfoDictionary {
 
+  /// A Boolean value that indicates whether this executable was an optimized (not debug) build.
+  #if DEBUG
+  let isRelease = false
```
I'm thinking this should be renamed `isDebug`. This is tied to the `Preprocessor Macros` Xcode setting. It is indicating the Debug Xcode build configuration was used.

low-batt commented on iina/InfoDictionary.swift:
```diff
@@ -44,14 +50,57 @@ struct InfoDictionary {
+  ///
+  /// This corresponds to the Xcode build configuration.
+  var buildType: BuildType {
+    guard let buildConfiguration = buildConfiguration else { return .nightly }
```
Will other developers be able to explain why this defaults to `.nightly`?

low-batt reviewed (commented):
I tested each build configuration and confirmed the logging issue is fixed. However, there is one issue left that I'm thinking needs to be addressed. I completely missed that the commit in the about window was a link because it does not follow the underscored blue text convention for links. As the other links in the about window do adhere to that convention this seems like something that should be fixed. Only the commit number should be linked, not the branch.

Fine to not use a link in the initial window. The about window is the primary source of information.

My two other comments are minor and not critical to address.

uiryuu committed:
Change isRelease to isDebug
